mountpoint-s3/src/inode.rs

```diff
@@ -131,55 +131,63 @@ impl Superblock {
-                        debug_assert!(false, "children should be forgotten before parents");
+        // Put inode removal in a block so we don't hold the lock on the inodes table longer than needed.
+        let (inode, parent) = {
+            let mut inodes = self.inner.inodes.write().unwrap();
+            match inodes.get(&ino) {
+                None => {
+                    debug_assert!(
+                        false,
+                        "forget should not be called on inode already removed from superblock"
+                    );
+                    error!("forget called on inode {ino} already removed from the superblock");
+                    return;
+                }
+                Some(inode) => {
+                    logging::record_name(inode.name());
+                    let new_lookup_count = inode.dec_lookup_count(n);
+                    if new_lookup_count == 0 {
+                        // Safe to remove, kernel no longer has a reference to it.
+                        trace!(ino, "removing inode from superblock");
+                        let inode = inodes.remove(&ino).expect("we just retrieved it");
+
+                        // Should be impossible for this to fail (VFS inodes reference their parent, so
+                        // children need to be freed first), but let's not crash in a `forget` function...
+                        let Some(parent) = inodes.get(&inode.parent()) else {
+                            debug_assert!(false, "children should be forgotten before parents");
+                            return;
+                        };
+                        (inode, parent.clone())
+                    } else {
                         return;
-                    };
-                    let mut parent_state = parent.inner.sync.write().unwrap();
-                    let InodeKindData::Directory {
-                        children,
-                        writing_children,
-                        ..
-                    } = &mut parent_state.kind_data
-                    else {
-                        unreachable!("parent is always a directory");
-                    };
-                    if let Some(child) = children.get(inode.name()) {
-                        // Don't accidentally remove a newer inode (e.g. remote shadowing local)
-                        if child.ino() == ino {
-                            children.remove(inode.name());
-                        }
                     }
-                    writing_children.remove(&ino);
-
-                    if let Ok(state) = inode.get_inode_state() {
-                        metrics::counter!(
-                            "metadata_cache.inode_forgotten_before_expiry",
-                            state.stat.is_valid().into(),
-                        );
-                    };
                 }
             }
+        };
+
+        let mut parent_state = parent.inner.sync.write().unwrap();
```

Comment:
Is it possible for the parent of the inode to change from the moment you capture it in the inner block?

Reply:
It is possible, but I think that is true even when we locked the `inodes` because parent is actually a shared reference.